torchaudio/filtering.cpp

```diff
@@ -0,0 +1,80 @@
+#include <torch/extension.h>
+
+//#include <algorithm>
+//#include <cstdint>
+//#include <stdexcept>
+//#include <vector>
+
+namespace torch {
+namespace audio {
+
+  // N.B. only handles floating point right now
+  void diff_eq(
+    at::Tensor const & input_waveform,
+    at::Tensor& output_waveform,
+    at::Tensor const & a_coeffs,
+    at::Tensor const & b_coeffs
+  ) {
+
+    // assumes waveform is normalized between 1 and -1
+
+    int64_t n_channels = input_waveform.size(0);
+    int64_t n_frames = input_waveform.size(1);
+
+    assert(output_waveform.size(0) == n_channels);
+    assert(output_waveform.size(1) == n_frames);
+
+    auto input_accessor = input_waveform.accessor<float,2>();
+    auto output_accessor = output_waveform.accessor<float,2>();    
+    auto a_coeffs_accessor = a_coeffs.accessor<float,1>();
+    auto b_coeffs_accessor = b_coeffs.accessor<float,1>();
+
+    int n_order = a_coeffs.size(0); // n'th order - 1 filter
+    assert(a_coeffs.size(0) == b_coeffs.size(0));
+
+    for (int64_t i_channel = 0; i_channel < n_channels; ++i_channel) {
```

Comment:
I'm not sure there is necessarily a faster way.  Would you use tensors to slice all channels at each frame?  Would that be much faster?

Reply:
Slicing would be great, yes. I'd expect this to work faster on GPU.

Reply:
@yf225 -- We want to implement a transformation `lfilter` that could "almost" be implemented by convolutions. We can't because of the way the terms depend on each other. @engineerchuan suggests implementing it in C++ (see here, in `torchaudio/filtering.cpp`), since the for-loop in time is much faster and comparable to scipy in speed on CPU. We shouldn't need a for-loop over the channels though. Thoughts on things to be careful about here?

Reply:
Since the computation for each channel doesn't depend on each other, we might be able to use `at::parallel_for` (https://github.com/pytorch/pytorch/blob/cc61af3c3d8ca8b46f7234383513b5166e10150c/aten/src/ATen/Parallel.h#L48) to speed up on CPU, and `thrust::for_each` (or a custom CUDA kernel launch) to speed up on GPU.